### ISSUE TITLE
[Issue] Improve integration with Mopidy-Iris

### ISSUE THREAD

**Issue by tehkillerbee** [tehkillerbee/mopidy-tidal#61]
After testing Mopidy-Tidal together with Iris, it is clear that there are improvements needed as outlined [here](https://github.com/jaedb/Iris/issues/363), 

Essentially, the following tasks must be fixed to give a better user experience:

- [ ] Speed up album art display, eg with additional local caching. This is especially visible when running on a RPi3. Sometimes a refresh is needed before the album arts are displayed. Cold startup results in a very long delay before album arts are displayed, likely since they have to be cached first.
- [ ] When browsing using the Iris "Playlists"/"Artists" shortcuts, album art is not shown. When browsing using "Tidal/My Playlists", "Tidal/My Artists", album art is displayed. It looks like the album art is never requested.
- [ ] Better playlist caching in general, as described in #60 
- [ ] Playlist editing

Feel free to add more comments in case you have discovered other issues.

**Comment by BlackLight** [comment]
Can we do a quick review of the tasks on this issue? After the PRs merged (or submitted) in the past couple of months, it looks to me like most of these points have been addressed.

**Comment by tehkillerbee** [comment]
@BlackLight I have tested and updated the issue

**Comment by 2e0byo** [comment]
I don't see any album art when I go to 'playlists', but I do when going to browse/tidal/my playlists, when I have a cold cache (i.e. `rm -r ~/.cache/mopidy`).  @tehkillerbee did you test against a cold cache?

**Comment by tehkillerbee** [comment]
@2e0byo Yes, you are right. This issue occurs when browsing Iris - Playlists, Artists shortcut but not when browsing the Tidal - My Playlists, My Albums, My Artists shortcuts.

**Comment by Alex45463** [comment]
hello I've just started using Mopidy-Tidal and Iris and I found some issues that I wanted to solve. 
This said, I want to contribute to this project

**Comment by Alex45463** [comment]
@tehkillerbee i have some issue to loading my liked tracks from tidal and some synchronization issues with Iris+Tidal.

The first issue happens when i just try to see my liked songs.
I have 700+ songs and that cause a long fetch for the URIs and seems to locks the thread of mpd until it finished to fetch every song. 

Idk if it's mopidy-tidal or tidal-api. 
An help is grateful to find the issue and to solve it.


Said that, my ideal setup would be a streaming server that allows me to play from tidal and local (ideally extendible with plugins), allowing to control the device where the song is played and switch to one another. (Like Spotify)

**Comment by BlackLight** [comment]
> The first issue happens when i just try to see my liked songs.
I have 700+ songs and that cause a long fetch for the URIs and seems to locks the thread of mpd until it finished to fetch every song.

I have never tried to access my liked songs directly because I have ~10k of them and I'm pretty sure that that would cause mopidy to freeze for hours.

That being said, mopidy freezing while it runs the lookup isn't something we can do much about, given the mostly single-threaded nature of mopidy. The same goes for the initial lookup: unfortunately mopidy performs lookups with one track at the time, so bulk API calls to speed up things aren't really an option. However, after the first lookup all the tracks should be cached, and successive lookups should be considerably faster. Can you confirm that this is indeed the case?

**Comment by Alex45463** [comment]
> I have never tried to access my liked songs directly because I have ~10k of them and I'm pretty sure that that would cause mopidy to freeze for hours.
> 
> That being said, mopidy freezing while it runs the lookup isn't something we can do much about, given the mostly single-threaded nature of mopidy. The same goes for the initial lookup: unfortunately mopidy performs lookups with one track at the time, so bulk API calls to speed up things aren't really an option. However, after the first lookup all the tracks should be cached, and successive lookups should be considerably faster. Can you confirm that this is indeed the case?

Apparently i cannot even lookup for all the songs even if i wait. (I also tried to access my Tracks to browse > Tidal). It could be a timeout from either Iris or Mopidy that is preventing the loading of them. At the moment my internet speed is pretty slow and that could be a factor resulting in these timeouts, but i can play Tidal songs thru my Playlists, my albums or artits.

**Comment by BlackLight** [comment]
You can probably take a look at the mopidy logs when it hangs. As long as you see `Lookup URIs ` lines, it means that the backend is loading the tracks. It's just that it may take a long time, and in the meantime it may be a good idea to avoid refreshes because they could trigger even more lookups.

I have personally never tried to load ~5000 tracks in the main playlist, but loading my largest playlist (a little under 3000 tracks) on a RPi4 takes ~10/15 minutes on the first lookup. After the data is cached, it takes ~30 seconds to load the playlist. If you're on a slow connection and you're loading twice the number of tracks in my largest playlist, then you may easily expect the initial lookup and caching to take ~30 minutes. Subsequent lookups should take much less though.
